### PR TITLE
feat: Make AppKernel with KernelTestCase compatible

### LINKED ISSUE #39: Compatibility with FrameworkBundle Test Traits
I just noticed that the `BaseBundleTestCase`  class is not compatible with the FrameworkBundle AssertionTraits.

In the trait `MailerAssertionsTrait` the [`getContainer`](https://github.com/symfony/symfony/blob/c8f37996f648bcd18c8fe0cb60cba0e6a0b7c6d9/src/Symfony/Bundle/FrameworkBundle/Test/MailerAssertionsTrait.php#L121) function is statically called, cause it is expected that the trait will be used in the `KernelTestCase` which have all the kernel and functions static.

Unfortunately the `BaseBundleTestCase` uses [non-static variables and functions](https://github.com/chapterjason/symfony-bundle-test/blob/bdfa9c7f0b3beba317745c7026e05b8614a8b6dc/src/BaseBundleTestCase.php#L49) and causes errors like these:

```
Error : Non-static method Nyholm\BundleTest\BaseBundleTestCase::getContainer() cannot be called statically
 /[...]/vendor/symfony/framework-bundle/Test/MailerAssertionsTrait.php:121
 /[...]/vendor/symfony/framework-bundle/Test/MailerAssertionsTrait.php:25
 /[...]/Tests/ExampleTest.php:159
```

I see three possible solutions:

1. We change the functions and variables to static, which would be a breaking change.
2. We serve custom traits with the same functionality which uses `$this`, but this also means more maintenance.
3. Combination of 1 and 2. We first serve custom traits to avoid a breaking change and change to static in version 2.0.

// EDIT
Also noticed that the KernelTextCase returns the `test.service_container` to allow access to private services.

### PR DESCRIPTION
**Pull request by chapterjason**
- Fix AppKernel constructor to be compatible with the KernelTestCase
- Add tests for usage with KernelTestCase
- Remove BaseBundleTestCase
- Update example in Readme

fixes: #39

### CODE REVIEW

**Comment by chapterjason** [comment]
@Nyholm I don't know how to deal with the failing tests:

```
PHP Fatal error:  Declaration of Symfony\Bundle\FrameworkBundle\Test\KernelTestCase::tearDown() must be compatible with PHPUnit\Framework\TestCase::tearDown(): void in /home/runner/work/symfony-bundle-test/symfony-bundle-test/vendor/symfony/framework-bundle/Test/KernelTestCase.php on line 24
```

Just found the following: https://github.com/symfony/symfony/issues/30071

**Comment by chapterjason** [comment]
@Nyholm I found a solution, hope we can use it like that.

File: src/AppKernel.php [hunk 1]
```diff
@@ -46,14 +46,23 @@ class AppKernel extends Kernel
...
-    public function __construct($cachePrefix)
+    public function __construct(/*string $environment, bool $debug*/)
     {
-        parent::__construct($cachePrefix, true);
-        $this->cachePrefix = $cachePrefix;
-        $this->addBundle(FrameworkBundle::class);
+        $args = \func_get_args();
+
+        if (1 === \func_num_args()) {
+            trigger_deprecation('nyholm/symfony-bundle-test', '1.9', 'The signature of the "%s($cachePrefix)" constructor is deprecated, use the constructor with 2 arguments: "string $environment, bool $debug".', self::class);
```
Comment (chapterjason) [inline src/AppKernel.php]: I also thought about just using `var_dump` or something else, cause the trigger_deprecation doesn't print anything out so it will never be visible for the developer without a prober error handler registered.

**Comment by Nyholm** [comment]
Awesome. Thank you. 

Could you update this PR after bumping composer.json. See https://github.com/SymfonyTest/symfony-bundle-test/pull/45#issuecomment-890694242

**Comment by chapterjason** [comment]
@Nyholm Yes, I have done some other changes too, I would like do this at the end because of the renamed tests. :)

**Comment by chapterjason** [comment]
@Nyholm Also ready now, without the breaking change layer. Ready for a major.

File: Readme.md [hunk 2]
```diff
@@ -52,17 +65,17 @@ class BundleInitializationTest extends BaseBundleTestCase
...
         // Boot the kernel as normal ...
-        $this->bootKernel();
+        $kernel = self::bootKernel([
+            'bundles' => [
+                // Add some other bundles we depend on
+                OtherBundle::class,
+            ],
+            'configFiles' => [
+                // Add some configuration
+                __DIR__.'/config.yml',
```
Comment (Nyholm) [inline Readme.md]: Can we keep `$kernel->addConfigFile(__DIR__.'/config.yml');`?

File: src/AppKernel.php [hunk 3]
```diff
@@ -179,6 +180,11 @@ public function loadRoutes(LoaderInterface $loader)
         return $routes->build();
     }
 
+    public function setCachePrefix($cachePrefix)
```
Comment (Nyholm) [inline src/AppKernel.php]: Why is this needed?

File: src/AppKernel.php [hunk 4]
```diff
@@ -208,4 +214,35 @@ public function setRoutingFile($routingFile)
     {
         $this->routingFile = $routingFile;
     }
+
+    public function handleOptions(array $options)
```
Comment (Nyholm) [inline src/AppKernel.php]: This is a new feature you snuck in. Why is it better to have an array of values rather than calling the underlying methods?

File: tests/Functional/BundleConfigurationTest.php [hunk 5]
```diff
@@ -22,12 +32,17 @@ public function provideBundleWithDifferentConfigurationFormats()
...
+                function (ContainerBuilder $container) {
+                    $container->loadFromExtension(
+                        'configuration',
+                        [
+                            'foo' => 'val1',
+                            'bar' => ['val2', 'val3'],
+                        ]
+                    );
+                },
+            ],
```
Comment (Nyholm) [inline tests/Functional/BundleConfigurationTest.php]: Is this change just CS?

I would prefer that PRs does not include CS changes on unrelated code.

**Review by Nyholm — CHANGES_REQUESTED** [review]
Thank you for working on this

File: Readme.md (hunk 2 above)
Comment (chapterjason) [reply Readme.md]: Sure. Have a look in the other comment.

File: src/AppKernel.php (hunk 3 above)
Comment (chapterjason) [reply src/AppKernel.php]: Before, it was just used the first argument from the constructor, I think we could also remove this.
I think there is not a need to be able to configure this.

File: tests/Functional/BundleConfigurationTest.php (hunk 5 above)
Comment (chapterjason) [reply tests/Functional/BundleConfigurationTest.php]: Was unintended, I will revert this here.

File: src/AppKernel.php (hunk 4 above)
Comment (chapterjason) [reply src/AppKernel.php]: 😄 Yeah.

I liked the idea of the options in the `KernelTestCase`, you are right, the methods are better.

The reason why I introduced this is: You only have the possibility to change the config by override the `createKernel` this limits it to a configuration per file and not per test. (If you don't want to boot it yourself and just want to use `static:bootKernel`.)

By just call the `handleOptions` in the `createKernel` you can actually passthrough everything before the kernel is booted and the container is compiled.

Another Idea, that just came in my mind, we could just create a option like `config` which must be a callable:

```php
static::bootKernel(['config' => static function(AppKernel $kernel){
    $kernel->setRoutingFile(__DIR__.'/Fixtures/routes.yml');
}]);
```

Then we can benefit from the methods and save some lines, cause the `KernelTestCase` makes the booting process.

File: src/BaseBundleTestCase.php [hunk 6]
```diff
@@ -1,113 +0,0 @@
...
-
-use PHPUnit\Framework\TestCase;
-use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
-use Symfony\Component\DependencyInjection\ResettableContainerInterface;
-use Symfony\Component\HttpKernel\Kernel;
-
-/**
- * @author Tobias Nyholm <tobias.nyholm@gmail.com>
- */
-abstract class BaseBundleTestCase extends TestCase
```
Comment (Nyholm) [inline src/BaseBundleTestCase.php:13]: This is great. Im happy to remove this class.

**Review by Nyholm — APPROVED** [review]
Thank you
